checker/src/main/java/org/checkerframework/checker/determinism/DeterminismVisitor.java

```diff
@@ -464,66 +466,96 @@ private void checkForDetConditional(ExpressionTree conditionalExpression) {
         }
     }
 
-    /**
-     * Reports an error if {@code @PolyDet("up")}, {@code @PolyDet("down")},
-     * {@code @PolyDet("use")}, or {@code @PolyDet("upDet")} is written on a formal parameter or a
-     * return type and none of the formal parameters or the receiver has the type {@code @PolyDet}.
-     */
     @Override
     public Void visitMethod(MethodTree node, Void p) {
-        HashSet<AnnotationMirror> polyAnnotations = new HashSet<>();
-        VariableTree receiverParam = node.getReceiverParameter();
-        if (receiverParam != null) {
-            polyAnnotations.add(
+        checkMethodSignatureForPolyQuals(node);
+        return super.visitMethod(node, p);
+    }
+
+    /**
+     * Reports an error if polymorphic qualifier that does not affect instantiation is used in a
+     * method signature without a polymorphic qualifier that does affect instantiation. Without at
+     * least one qualifier that affects instantiation, the checker cannot resolve the other
+     * polymorphic qualifiers.
+     *
+     * <p>Polymorphic qualifiers that do not affect instantiation: any polymorphic qualifier on the
+     * return type and any of these annotations on a parameter or receiver type:
+     * {@code @PolyDet("up")}, {@code @PolyDet("down")}, {@code @PolyDet("upDet")}, and
+     * {@code @PolyDet("use")}.
+     *
+     * <p>Polymorphic qualifiers that do affect instantiation: {@code @PolyDet} and
+     * {@code @PolyDet("noOrderNonDet")} on a parameter or receiver type.
+     */
+    private void checkMethodSignatureForPolyQuals(MethodTree methodTree) {
+        // Errors that should be issued if @PolyDet or @PolyDet("noOrderNonDet") are not found.
+        List<Pair<Result, Tree>> errors = new ArrayList<>();
+        VariableTree receiver = methodTree.getReceiverParameter();
+        if (receiver != null) {
+            AnnotationMirror anno =
                     atypeFactory
-                            .getAnnotatedType(receiverParam)
-                            .getAnnotationInHierarchy(atypeFactory.NONDET));
-        } else if (!node.getModifiers().getFlags().contains(Modifier.STATIC)) {
-            polyAnnotations.add(atypeFactory.POLYDET);
+                            .getAnnotatedType(receiver)
+                            .getAnnotationInHierarchy(atypeFactory.NONDET);
+            if (addPolyDetError(anno, receiver, errors)) {
+                return; // found @PolyDet or @PolyDet("noOrderNonDet")
+            }
+        } else if (!ElementUtils.isStatic(TreeUtils.elementFromDeclaration(methodTree))) {
+            // This is an instance method without explicit "this" parameter.
+            // TODO: this is what this method used to do, but I don't understand why.
```

Comment:
I re-wrote `visitMethod` so that it was easier to understand.  But I still don't understand why non-static methods without an explicit receiver aren't checked.  @t-rasmud can you explain?

Reply:
Thanks! Non-static methods without an explicit receiver should be checked. I fixed this in 5ee382b407d17262036806ad95e61c534bdf6f1b.
The `<init>` constructor is a non-static method whose receiver type is `null` and my latest changes bypass the check just for this method.